### PR TITLE
wip, added tactic to narrow state, tactic to register new lemma

### PR DESCRIPTION
**Pull request by Scikud**
WIP added tactics to narrow state to conjecture, register new lemma given prf expr.

### CODE REVIEW

File: src/conjecture_tactics.lean [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    | nil          := undefined
+    end
+  | (some new) := return (mk_str_name old.get_prefix new.to_string)
+end
+
+
+/--  
+  return value ⟨nm, ts⟩ satisfies that `nm` is the name of a new declaration whose value is the proof `conj_pf` of the conjecture and that `ts` contains an environment which has `nm` registered as a declaration.
+-/
+meta def tac₂ (conj_pf : expr) (narrowed_ts : tactic_state) : tactic (name × tactic_state) := do {
```
Comment (Scikud) [inline src/conjecture_tactics.lean:50]: @jesse-michael-han how does this second tactic look?

File: src/conjecture_tactics.lean [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+/--
+  `tac₃` should take the old_ts emitted by `tac₁`, looks up the declaration `nm` in `new_ts`'s environment, and uses it to prove the open conjecture which is the top-level goal of `old_ts`, and then returns `old_ts` after this modification. this represents the rest of the proof search after the conjecture been "re-injected".
+-/
+meta def tac₃ (old_ts : tactic_state) (nm : name) (new_ts : tactic_state) : tactic tactic_state := do {
+  old_gs ← tactic.get_goals,
+  tactic.write new_ts,
+  modified_env ← tactic.get_env, 
+  d ← modified_env.get nm,
+
+  sorry
```
Comment (Scikud) [inline src/conjecture_tactics.lean:80]: @jesse-michael-han I'm a little unclear on how we use the decl from the old tactic state to prove the open conjecture. Mind clarifying a little :) ?

File: src/conjecture_tactics.lean (hunk 1 above)
Comment (jesse-michael-han) [reply src/conjecture_tactics.lean:50]: tentatively LGTM, can you try testing it with something like this?

```lean
/--
  should succeed iff tac₂ meets its spec
-/
meta def test_tac₂ : tactic unit := do {
  let pf_term : expr := `(trivial),
  ts ← tactic.read,
  ⟨nm ,new_ts⟩ ← tac₂ pf_term ts,
  tactic.write new_ts,
  env ← tactic.get_env,
  decl ← env.get nm, -- succeeds iff `nm` is inside `env`
  pure ()
}

```

File: src/conjecture_tactics.lean (hunk 2 above)
Comment (jesse-michael-han) [reply src/conjecture_tactics.lean:80]: try:
 - pattern-matching on the declaration `d` to get the proof term expr `p`, then
 - using `tactic.exact p` to close the first goal

File: src/conjecture_tactics.lean [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    let thunk := fn.get_or_else (λ _, format! "exception"),
+    tactic.fail format! "{thunk ()}"
+  }
+  end),
+  pure $ prod.mk decl.to_name modified_ts
+}
+
+/--
+  `tac₃` should take the old_ts emitted by `tac₁`, looks up the declaration `nm` in `new_ts`'s environment, and uses it to prove the open conjecture which is the top-level goal of `old_ts`, and then returns `old_ts` after this modification. this represents the rest of the proof search after the conjecture been "re-injected".
+-/
```
Comment (Scikud) [inline src/conjecture_tactics.lean:73]: @jesse-michael-han lmk if this final tactic looks good to you

**Pull request closed without merging**